File: public/packages/typescript/caching/abstractions/src/input-failures.ts
// -----------------------------------------------------------------------
// SPDX-License-Identifier: Apache-2.0
// Copyright (c) DCSV
// -----------------------------------------------------------------------

import { TK } from "@d2/i18n-keys";
import { inputError, validationFailed, type D2Result } from "@d2/result";

/**
 * Pre-built `D2Result` input-failure responses for cache impls.
 * Keeps the cache surface pure errors-as-values instead of mixing in
 * throws for caller mistakes.
 *
 * Constructors / DI registration still throw — that is a different
 * lifecycle concern from per-call input validation.
 *
 * Twin of .NET `D2.Shared.Caching.InputFailures`.
 */
function requiredImpl<T = void>(paramName: string): D2Result<T> {
  return validationFailed<T>({
    inputErrors: [inputError(paramName, [TK.common.errors.NOT_NULL_VIOLATION])],
  });
}

function invalidImpl<T = void>(paramName: string): D2Result<T> {
  return validationFailed<T>({
    inputErrors: [inputError(paramName, [TK.common.errors.VALIDATION_FAILED])],
  });
}

export const InputFailures = {
  /**
   * Builds a `D2Result` / `D2Result<T>` input failure for a missing
   * (null / empty / falsey) required parameter.
   *
   * @param paramName - Parameter name (use a string literal matching
   *   the call-site argument name).
   */
  required: requiredImpl as {
    (paramName: string): D2Result;
    <T>(paramName: string): D2Result<T>;
  },

  /**
   * Builds a `D2Result` / `D2Result<T>` input failure for a present but
   * invalid parameter value (range, non-finite, non-safe-integer, etc.).
   * Does **not** use `NOT_NULL_VIOLATION` — the value is present.
   *
   * @param paramName - Parameter name (use a string literal matching
   *   the call-site argument name).
   */
  invalid: invalidImpl as {
    (paramName: string): D2Result;
    <T>(paramName: string): D2Result<T>;
  },
};

File: public/packages/dotnet/caching/abstractions/InputFailures.cs
// -----------------------------------------------------------------------
// <copyright file="InputFailures.cs" company="DCSV">
// Copyright (c) DCSV. Licensed under the Apache License, Version 2.0.
// </copyright>
// -----------------------------------------------------------------------

namespace DcsvIo.D2.Caching;

using DcsvIo.D2.I18n;
using DcsvIo.D2.Result;

/// <summary>
/// Pre-built <see cref="D2Result"/> input-failure responses for cache impls.
/// Keeps the cache surface pure errors-as-values instead of mixing in
/// <see cref="ArgumentException"/> throws for caller mistakes.
/// </summary>
/// <remarks>
/// Constructors still throw — DI / construction failure is a different
/// lifecycle concern from per-call input validation.
/// </remarks>
public static class InputFailures
{
    /// <summary>
    /// Builds a <see cref="D2Result{TData}"/> input failure for a missing
    /// (null or empty) required parameter.
    /// </summary>
    /// <typeparam name="T">Wrapped result type.</typeparam>
    /// <param name="paramName">Parameter name (use <c>nameof</c>).</param>
    public static D2Result<T> Required<T>(string paramName)
        => D2Result<T>.ValidationFailed(
            inputErrors: [new InputError(paramName, [TK.Common.Errors.NOT_NULL_VIOLATION])]);

    /// <summary>
    /// Builds a non-generic <see cref="D2Result"/> input failure for a
    /// missing (null or empty) required parameter.
    /// </summary>
    /// <param name="paramName">Parameter name (use <c>nameof</c>).</param>
    public static D2Result Required(string paramName)
        => D2Result.ValidationFailed(
            inputErrors: [new InputError(paramName, [TK.Common.Errors.NOT_NULL_VIOLATION])]);

    /// <summary>
    /// Builds a <see cref="D2Result{TData}"/> input failure for a present but
    /// invalid parameter value (range, non-positive, etc.). Does not use
    /// <see cref="TK.Common.Errors.NOT_NULL_VIOLATION"/>.
    /// </summary>
    /// <typeparam name="T">Wrapped result type.</typeparam>
    /// <param name="paramName">Parameter name (use <c>nameof</c>).</param>
    public static D2Result<T> Invalid<T>(string paramName)
        => D2Result<T>.ValidationFailed(
            inputErrors: [new InputError(paramName, [TK.Common.Errors.VALIDATION_FAILED])]);

    /// <summary>
    /// Builds a non-generic <see cref="D2Result"/> input failure for a present
    /// but invalid parameter value (range, non-positive, etc.).
    /// </summary>
    /// <param name="paramName">Parameter name (use <c>nameof</c>).</param>
    public static D2Result Invalid(string paramName)
        => D2Result.ValidationFailed(
            inputErrors: [new InputError(paramName, [TK.Common.Errors.VALIDATION_FAILED])]);
}
